src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java

```diff
@@ -111,32 +111,44 @@ public ModelingSubmission getLockedModelingSubmissionWithoutResult(ModelingExerc
+     * transaction here as it is making problems when initially loading the calculation engine and assessing all submissions automatically: we would get an sql exception
+     * "Connection is read-only" from hibernate when saving the result in CompassService#assessAutomatically.
      *
      * @param modelingExercise the modeling exercise for which we want to get a modeling submission without result
      * @return a modeling submission without any result
      */
-    @Transactional(readOnly = true)
+    @Transactional
     public Optional<ModelingSubmission> getModelingSubmissionWithoutResult(ModelingExercise modelingExercise) {
-        // ask Compass for optimal submission to assess if diagram type is supported
+        // if the diagram type is supported by Compass, ask Compass for optimal (i.e. most knowledge gain for automatic assessments) submissions to assess next
         if (compassService.isSupported(modelingExercise.getDiagramType())) {
-            Set<Long> optimalModelSubmissions = compassService.getModelsWaitingForAssessment(modelingExercise.getId());
-            if (!optimalModelSubmissions.isEmpty()) {
-                // TODO CZ: think about how to handle canceled assessments with Compass as I do not want to receive the same submission again, if I canceled the assessment
-                return modelingSubmissionRepository.findById(optimalModelSubmissions.iterator().next());
+            List<Long> modelsWaitingForAssessment = compassService.getModelsWaitingForAssessment(modelingExercise.getId());
+
+            // shuffle the model list to prevent that the user gets the same submission again after canceling an assessment
+            Collections.shuffle(modelsWaitingForAssessment);
+
+            for (Long submissionId : modelsWaitingForAssessment) {
+                Optional<ModelingSubmission> submission = modelingSubmissionRepository.findByIdWithEagerResultAndFeedback(submissionId);
+                if (submission.isPresent()) {
+                    return submission;
+                }
+                else {
+                    compassService.removeModelWaitingForAssessment(modelingExercise.getId(), submissionId);
+                }
             }
         }
 
-        // otherwise return a random submission that is not assessed or an empty optional
-        Random r = new Random();
-        List<ModelingSubmission> submissionsWithoutResult = participationService.findByExerciseIdWithEagerSubmittedSubmissionsWithoutResults(modelingExercise.getId()).stream()
-                .map(Participation::findLatestModelingSubmission).filter(Optional::isPresent).map(Optional::get).collect(Collectors.toList());
+        // otherwise return a random submission that is not manually assessed or an empty optional if there is none
+        List<ModelingSubmission> submissionsWithoutResult = participationService.findByExerciseIdWithEagerSubmittedSubmissionsWithoutManualResults(modelingExercise.getId())
```

Comment:
It might be worth it to do this directly in SQL.
I have done something similar here: https://github.com/ls1intum/Artemis/blob/develop/src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java#L35

Reply:
You are right. I am not entirely if this is worth it, though. Normally, Compass should already return a submission that needs an assessment and this query is just the fallback. Additionally, the query would be even more complex than it already is and it would be kind of complex to test, if it behaves as it should be in every scenario. 
For these reasons I would propose to keep it as it is for now. What do you think @thilo-behnke @krusche?

Reply:
Given your explanation I think it's ok to keep it this way :+1: